File: recipes/subms-stats/rust/src/bootstrap.rs
//! Bootstrap confidence intervals. Behind the `bootstrap` Cargo
//! feature (on by default).
//!
//! Use this when reporting a percentile to know how WIDE the
//! confidence interval is - "p99 = 230us +- 50us" is meaningful;
//! "p99 = 230us" alone is missing the noise floor.

use crate::percentiles::percentile;

/// Bootstrap a percentile estimate with `iters` resamples, returning
/// the lower + upper bounds of the `confidence`-level interval.
/// `confidence` in `[0.0, 1.0]`; typical value `0.95`. Uses a
/// deterministic LCG (`SubMsLcg`-compatible constants) so results
/// are reproducible across runs given the same `seed`.
///
/// Cost is `O(iters * n)`; defaults that work well in practice:
/// `iters = 200..1000`, `confidence = 0.95`.
pub fn bootstrap_percentile_ci(
    samples: &[u64],
    q: f64,
    iters: usize,
    confidence: f64,
    seed: u64,
) -> (u64, u64) {
    if samples.is_empty() || iters == 0 {
        return (0, 0);
    }
    let mut state = seed | 1;
    let mut estimates = Vec::with_capacity(iters);
    for _ in 0..iters {
        let mut resample = Vec::with_capacity(samples.len());
        for _ in 0..samples.len() {
            state = state
                .wrapping_mul(6364136223846793005)
                .wrapping_add(1442695040888963407);
            let idx = (state as usize) % samples.len();
            resample.push(samples[idx]);
        }
        resample.sort_unstable();
        estimates.push(percentile(&resample, q));
    }
    estimates.sort_unstable();
    let lo_q = (1.0 - confidence) / 2.0;
    let hi_q = 1.0 - lo_q;
    (percentile(&estimates, lo_q), percentile(&estimates, hi_q))
}

#[cfg(test)]
#[path = "bootstrap_tests.rs"]
mod tests;

File: recipes/subms-stats/java/src/main/java/com/submillisecond/recipes/stats/Bootstrap.java
package com.submillisecond.recipes.stats;

import java.util.Arrays;

/**
 * Bootstrap confidence intervals for percentiles. Reproducible
 * resampling via a deterministic LCG.
 *
 * <p>Use when reporting a percentile to know how WIDE the confidence
 * interval is - "p99 = 230us +- 50us" is meaningful; "p99 = 230us"
 * alone is missing the noise floor.
 *
 * <p>Byte-equivalent to {@code subms_stats::bootstrap} in the Rust
 * sibling.
 */
public final class Bootstrap {
    private Bootstrap() {}

    /**
     * Result of a bootstrap CI: the lower and upper bounds of the
     * confidence interval, in nanoseconds.
     */
    public record CI(long lo, long hi) {}

    /**
     * Bootstrap a percentile estimate with {@code iters} resamples,
     * returning the lower + upper bounds of the {@code confidence}-
     * level interval. {@code confidence} in {@code [0.0, 1.0]};
     * typical value {@code 0.95}.
     *
     * <p>Cost is {@code O(iters * n)}; sensible defaults are
     * {@code iters = 200..1000}, {@code confidence = 0.95}.
     */
    public static CI bootstrapPercentileCi(long[] samples, double q, int iters, double confidence, long seed) {
        if (samples.length == 0 || iters == 0) return new CI(0L, 0L);
        long state = seed | 1L;
        long[] estimates = new long[iters];
        long[] resample = new long[samples.length];
        for (int it = 0; it < iters; it++) {
            for (int i = 0; i < samples.length; i++) {
                state = state * 6364136223846793005L + 1442695040888963407L;
                long idx = state >>> 1;
                resample[i] = samples[(int) (idx % samples.length)];
            }
            long[] sorted = resample.clone();
            Arrays.sort(sorted);
            estimates[it] = Percentiles.percentile(sorted, q);
        }
        Arrays.sort(estimates);
        double loQ = (1.0 - confidence) / 2.0;
        double hiQ = 1.0 - loQ;
        return new CI(
                Percentiles.percentile(estimates, loQ),
                Percentiles.percentile(estimates, hiQ));
    }
}
